src/Mayaqua/Encrypt.c

```diff
@@ -4587,135 +4587,176 @@ void AesFreeKey(AES_KEY_VALUE *k)
+
+	if (ret != 1)
 	{
-		is_intel_aes_supported = true;
+		ERR_print_errors_fp(stderr);
+		return;
 	}
-	else
+
+	// Finalize the encryption
+	int len = 0;
+	ret = EVP_EncryptFinal_ex(ctx, ciphertext + ciphertext_len, &len);
+
+	if (ret != 1)
 	{
-		is_intel_aes_supported = false;
+		ERR_print_errors_fp(stderr);
+		return;
 	}
-#else	// USE_INTEL_AESNI_LIBRARY
-	is_intel_aes_supported = false;
-#endif	// USE_INTEL_AESNI_LIBRARY
-}
 
-// Disable the Intel AES-NI
-void DisableIntelAesAccel()
-{
-	is_intel_aes_supported = false;
+	ciphertext_len += len;
+
+	Copy(dest, ciphertext, ciphertext_len);
+
+	// Clean up
+	EVP_CIPHER_CTX_free(ctx);
 }
 
-#ifdef	USE_INTEL_AESNI_LIBRARY
-// Encrypt AES using the Intel AES-NI
-void AesEncryptWithIntel(void *dest, void *src, UINT size, AES_KEY_VALUE *k, void *ivec)
+// AES decryption
+void AesDecrypt(void *dest, void *src, UINT size, AES_KEY_VALUE *k, void *ivec)
 {
-	UCHAR ivec_copy[AES_IV_SIZE];
-
 	// Validate arguments
 	if (dest == NULL || src == NULL || size == 0 || k == NULL || ivec == NULL)
 	{
 		return;
 	}
 
+	// Create and initialize the context
+	EVP_CIPHER_CTX *ctx = EVP_CIPHER_CTX_new();
+
+	if (!ctx)
+	{
+		ERR_print_errors_fp(stderr);
+		return;
+	}
+
+	// Initialize the decryption operation
+	UCHAR ivec_copy[AES_IV_SIZE];
 	Copy(ivec_copy, ivec, AES_IV_SIZE);
 
+	int ret = 0;
+
 	switch (k->KeySize)
 	{
 	case 16:
-		intel_AES_enc128_CBC(src, dest, k->KeyValue, (size / AES_IV_SIZE), ivec_copy);
+		ret = EVP_DecryptInit_ex(ctx, EVP_aes_128_cbc(), NULL, k->KeyValue, ivec_copy);
 		break;
 
 	case 24:
-		intel_AES_enc192_CBC(src, dest, k->KeyValue, (size / AES_IV_SIZE), ivec_copy);
+		ret = EVP_DecryptInit_ex(ctx, EVP_aes_192_cbc(), NULL, k->KeyValue, ivec_copy);
 		break;
 
 	case 32:
-		intel_AES_enc256_CBC(src, dest, k->KeyValue, (size / AES_IV_SIZE), ivec_copy);
+		ret = EVP_DecryptInit_ex(ctx, EVP_aes_256_cbc(), NULL, k->KeyValue, ivec_copy);
 		break;
 	}
-}
 
-// Decrypt AES using the Intel AES-NI
-void AesDecryptWithIntel(void *dest, void *src, UINT size, AES_KEY_VALUE *k, void *ivec)
-{
-	UCHAR ivec_copy[AES_IV_SIZE];
-
-	// Validate arguments
-	if (dest == NULL || src == NULL || size == 0 || k == NULL || ivec == NULL)
+	if (ret != 1)
 	{
+		ERR_print_errors_fp(stderr);
 		return;
 	}
 
-	Copy(ivec_copy, ivec, AES_IV_SIZE);
+	// Provide the message to be decrypted and obtain the plaintext output
+	UCHAR ciphertext[size];
+	Copy(ciphertext, src, size);
 
-	switch (k->KeySize)
+	UCHAR plaintext[size];
+	int plaintext_len = 0;
+
+	ret = EVP_DecryptUpdate(ctx, plaintext, &plaintext_len, ciphertext, size);
+
+	if (ret != 1)
 	{
-	case 16:
-		intel_AES_dec128_CBC(src, dest, k->KeyValue, (size / AES_IV_SIZE), ivec_copy);
-		break;
+		ERR_print_errors_fp(stderr);
+		return;
+	}
 
-	case 24:
-		intel_AES_dec192_CBC(src, dest, k->KeyValue, (size / AES_IV_SIZE), ivec_copy);
-		break;
+	// Finalize the decryption
+	int len = 0;
+	ret = EVP_DecryptFinal_ex(ctx, plaintext + plaintext_len, &len);

```

Comment:
If you remove the `EVP_EncryptFinal_ex` call as I suggested above, this call also should be removed. Otherwise you have to make sure the correct PKCS padding is present in the ciphertext when AesDecrypt is called --- which means adding it manually, because IPsec does not use PKCS padding.

Reply:
Using `EVP_CIPHER_CTX_set_padding()` seems to be the best solution at the moment.